### DISCUSSION TITLE
[Bug]: AIO_URL in configuration.json wrong

### DISCUSSION THREAD

**Discussion by distancesprinter** [nextcloud/all-in-one#1453] [Questions] [answered]
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on Github _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] Nextcloud Server **is** running on 64bit capable CPU, PHP and OS.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

Installed Nextcloud AIO Docker and specified my URL at setup.  Skipped domain validation.  I am using an Nginx reverse proxy.  Nextcloud is (mostly) working.  (I am struggling to get Collabora to work; seems like a websockets issue that I will investigate next.)

Noticed the URL in Settings/Administration/Overview for "Open Nextcloud AIO Interface" is using a domain different than what I specified when I installed Nextcloud.  That link doesn't work unless I copy the URL and fix the domain.  Then it works.

Researched and found that the configuration is in:
/var/lib/docker/volumes/nextcloud_aio_mastercontainer/_data/data/configuration.json

Opened that file, updated the AIO_URL, saved.  Then restarted the containers.  I was expecting this to fix the URL, but it is as if the AIO_URL was reverted back to the previous value when the containers came back up.  The URL is still wrong in Settings/Administration/Overview for Open Nextcloud AIO Interface.

Where is the system getting the AIO_URL?  How do I prevent it from reverting back to an internal URL that is not externally accessible or routable?  This seems like a bug.

### Steps to reproduce

1. Log in to Nextcloud with admin username; Settings/Administration/Overview for "Open Nextcloud AIO Interface" has wrong URL
2. Update AIO_URL in /var/lib/docker/volumes/nextcloud_aio_mastercontainer/_data/data/configuration.json
3. Restart containers

### Expected behavior

"Open Nextcloud AIO Interface" should have modified AIO_URL which would properly work when using the domain set in the Reverse Proxy.

### Installation method

Community Docker image

### Operating system

Debian/Ubuntu

### PHP engine version

PHP 8.0

### Web server

Apache (supported)

### Database engine version

PostgreSQL

### Is this bug present after an update or on a fresh install?

Fresh Nextcloud Server install

### What user-backends are you using?

- [X] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

```shell
{
    "system": {
        "memcache.local": "\\OC\\Memcache\\APCu",
        "apps_paths": [
            {
                "path": "\/var\/www\/html\/apps",
                "url": "\/apps",
                "writable": false
            },
            {
                "path": "\/var\/www\/html\/custom_apps",
                "url": "\/custom_apps",
                "writable": true
            }
        ],
        "memcache.distributed": "\\OC\\Memcache\\Redis",
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "password": "***REMOVED SENSITIVE VALUE***",
            "port": 6379
        },
        "overwritehost": "cloud.viverette.technology",
        "overwriteprotocol": "https",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "localhost",
            "cloud.viverette.technology"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "pgsql",
        "version": "24.0.7.1",
        "overwrite.cli.url": "https:\/\/cloud.viverette.technology\/",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "loglevel": "2",
        "log_type": "file",
        "logfile": "\/var\/www\/html\/data\/nextcloud.log",
        "log_rotate_size": "10485760",
        "log.condition": {
            "apps": [
                "admin_audit"
            ]
        },
        "preview_max_x": "2048",
        "preview_max_y": "2048",
        "jpeg_quality": "60",
        "enabledPreviewProviders": {
            "1": "OC\\Preview\\Image",
            "2": "OC\\Preview\\MarkDown",
            "3": "OC\\Preview\\MP3",
            "4": "OC\\Preview\\TXT",
            "5": "OC\\Preview\\OpenDocument",
            "6": "OC\\Preview\\Movie",
            "0": "OC\\Preview\\Imaginary"
        },
        "enable_previews": true,
        "upgrade.disable-web": true,
        "mail_smtpmode": "smtp",
        "trashbin_retention_obligation": "auto, 30",
        "versions_retention_obligation": "auto, 30",
        "activity_expire_days": "30",
        "simpleSignUpLink.shown": false,
        "share_folder": "\/Shared",
        "tempdirectory": "\/mnt\/ncdata\/tmp\/",
        "one-click-instance": true,
        "one-click-instance.user-limit": 100,
        "one-click-instance.link": "https:\/\/nextcloud.com\/all-in-one\/",
        "htaccess.RewriteBase": "\/",
        "files_external_allow_create_new_local": false,
        "trusted_proxies": "***REMOVED SENSITIVE VALUE***",
        "allow_local_remote_servers": true,
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "mail_sendmailmode": "smtp",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpauthtype": "PLAIN",
        "mail_smtpauth": 1,
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpsecure": "ssl",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "465",
        "default_phone_region": "US",
        "preview_imaginary_url": "http:\/\/nextcloud-aio-imaginary:9000"
    }
}
```

### List of activated Apps

```shell
Enabled:
  - accessibility: 1.10.0
  - activity: 2.16.0
  - admin_audit: 1.14.0
  - apporder: 0.15.0
  - calendar: 3.5.2
  - circles: 24.0.1
  - cloud_federation_api: 1.7.0
  - comments: 1.14.0
  - contacts: 4.2.2
  - contactsinteraction: 1.5.0
  - dashboard: 7.4.0
  - dav: 1.22.0
  - deck: 1.7.3
  - federatedfilesharing: 1.14.0
  - federation: 1.14.0
  - files: 1.19.0
  - files_antivirus: 4.0.0
  - files_fulltextsearch: 24.0.1
  - files_pdfviewer: 2.5.0
  - files_rightclick: 1.3.0
  - files_sharing: 1.16.2
  - files_trashbin: 1.14.0
  - files_versions: 1.17.0
  - files_videoplayer: 1.13.0
  - firstrunwizard: 2.13.0
  - fulltextsearch: 24.0.0
  - fulltextsearch_elasticsearch: 24.0.1
  - groupfolders: 12.0.2
  - logreader: 2.9.0
  - lookup_server_connector: 1.12.0
  - nextcloud-aio: 0.2.0
  - nextcloud_announcements: 1.13.0
  - notifications: 2.12.1
  - notify_push: 0.5.0
  - oauth2: 1.12.0
  - password_policy: 1.14.0
  - photos: 1.6.0
  - privacy: 1.8.0
  - provisioning_api: 1.14.0
  - recommendations: 1.3.0
  - richdocuments: 6.3.1
  - serverinfo: 1.14.0
  - settings: 1.6.0
  - sharebymail: 1.14.0
  - spreed: 14.0.6
  - support: 1.7.0
  - survey_client: 1.12.0
  - systemtags: 1.14.0
  - tasks: 0.14.5
  - text: 3.5.1
  - theming: 1.15.0
  - twofactor_backupcodes: 1.13.0
  - twofactor_totp: 6.4.1
  - user_status: 1.4.0
  - viewer: 1.8.0
  - weather_status: 1.4.0
  - workflowengine: 2.6.0
Disabled:
  - encryption
  - files_external
  - user_ldap
```

### Nextcloud Signing status

```shell
No errors have been found.
```

### Additional info

Firefox 107
Chrome 107

**Comment by szaimen** [comment]
Hi, did you see this? https://github.com/nextcloud/all-in-one#how-to-easily-log-in-to-the-aio-interface

**Reply by distancesprinter** [reply]
I did, but it's not working properly.  It states, "Note: You can change the domain/ip-address/port of the button by simply stopping the containers, visiting the AIO interface from the correct and desired domain/ip-address/port and clicking once on Start containers."

How does the AIO Master Container determine the URL?  My master container is behind multiple reverse proxies, so whatever mechanism it is using to determine the URL is probably incorrect.  That is why I want to specify it manually.  As previously mentioned, I had to skip domain validation when doing the install.

**Reply by szaimen** [reply]
> My master container is behind multiple reverse proxies, so whatever mechanism it is using to determine the URL is probably incorrect.

Yeah, that seems to be the case then. Why are you doing it so complicated?

**Reply by distancesprinter** [reply]
I set up these reverse proxies mainly as a learning exercise at first, but it's been working so well for all my other Docker containers that I've kept it in place.  Nextcloud is a very complicated collection of separate tools, but if it's going to be shipped for Docker, these are the sorts of things that need to be configurable, and I figure these are the sorts of things the team would want to iron out.  It's not uncommon to have Docker containers behind multiple layers of reverse proxies, load-balancers, etc.

**Comment by szaimen** [comment, marked as answer]
What about not making the AIO interace publicly available and instead only accessing it using a VPN or inside the home network directly? (That would also be recommended from my side anyways)...

**Reply by distancesprinter** [reply]
Yes, that obviously works fine.  It also works fine if I just copy the URL from the button and change the domain in my address bar.

I'm not sure why Nextcloud made the design decision to require an admin to access via the settings page.  I understand it improves security, but if you're going to do that, then you should provide an option to override or manually set the URL.  Especially since there exists an option to set the URL when you run the install.  So my purpose is to point out that that domain override is half-baked, because it doesn't affect or override whatever mechanism that determines the URL for the button.  I believe that it should.

I acknowledge this configuration is probably an edge case for most installs today, but because there's an option to manually set the domain, might as well get it right.  It might matter more in the future.

**Comment by Docmine17** [comment]
I solved it by putting that in my Caddyfile.:
reverse_proxy :port {
    header_up Host {http.request.host}
}
